components/resource_control/src/service.rs

```diff
@@ -139,6 +160,178 @@ impl ResourceManagerService {
+                Err(err) => {
+                    error!("failed to controller config"; "err" => ?err);
+                    let _ = GLOBAL_TIMER_HANDLE
+                        .delay(std::time::Instant::now() + RETRY_INTERVAL)
+                        .compat()
+                        .await;
+                    continue;
+                }
+            }
+        }
+    }
+
+    // upload ru metrics
+    pub async fn upload_ru_metrics(&self) {
+        let mut last_consumption_map: HashMap<String, Consumption> = HashMap::new();
+        // load controller config firstly
+        let config = self.load_controller_config().await.unwrap_or_default();
+
+        loop {
+            let background_groups: Vec<_> = self
+                .manager
+                .resource_groups
+                .iter()
+                .filter_map(|kv| {
+                    let g = kv.value();
+                    g.limiter.as_ref().map(|limiter| GroupStats {
+                        name: g.group.name.clone(),
+                        ru_quota: g.get_ru_quota() as f64,
+                        limiter: limiter.clone(),
+                        stats_per_sec: GroupStatistics::default(),
+                        expect_cost_rate: 0.0,
+                    })
+                })
+                .collect();
+            if background_groups.is_empty() {
+                let _ = GLOBAL_TIMER_HANDLE
+                    .delay(std::time::Instant::now() + BACKGROUND_RU_UPLOAD_DURATION)
+                    .compat()
+                    .await;
+                continue;
+            }
+
+            info!("[upload ru metrics] background_groups len is"; "background_groups" => background_groups.len());
+            let mut req = TokenBucketsRequest::default();
+            let all_reqs = req.mut_requests();
+            background_groups.iter().for_each(|g| {
+                let mut req = TokenBucketRequest::default();
+                req.set_resource_group_name(g.name.clone());
+                req.set_is_background(true);
+                // update statistics
+                let report_consumption = req.mut_consumption_since_last_request();
+                if let Some(last_consumption) = last_consumption_map.get_mut(g.name.as_str()) {
+                    let cpu_consumed = g
+                        .limiter
+                        .get_limiter(ResourceType::Cpu)
+                        .get_statistics()
+                        .total_consumed as f64;
+                    let io_consumed = g.limiter.get_limiter(ResourceType::Io).get_statistics();
+                    let read_bytes_consumed = io_consumed.read_consumed as f64;
+                    let write_bytes_consumed = io_consumed.write_consumed as f64;
+
+                    let read_total = config.cpu_ms_cost * cpu_consumed
+                        + config.read_cost_per_byte * read_bytes_consumed
+                        + config.read_per_batch_base_cost * DEFAULT_AVG_BATCH_PROPORTION;
```

Comment:
If you want to count `read/write_per_batch_base_cost`, you should count them either in the client or add a separate field to count the request count, add this default value makes no sense.